<a href="https://colab.research.google.com/github/sulthanalihsan/data-science-2026/blob/main/Pertemuan10_Muhamad_Sulthan_Al_Ihsan_250401020154.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Pertemuan 10 — Algoritma Klasifikasi (Bagian 2)
**Nama  :** Muhammad Sulthan Al Ihsan

**NIM   :** 250401020154

**Mata Kuliah:** Data Science —  S1 PJJ Informatika

**Kelas:** IF401

# **3.5 Implementasi dengan Scikit-Learn**

In [31]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.datasets import load_breast_cancer

# Memuat dataset contoh (misal: breast cancer) agar X dan y terdefinisi
data = load_breast_cancer()
X, y = data.data, data.target

# Pisahkan data latih dan uji, jaga proporsi kelas (stratify)
X_tr, X_te, y_tr, y_te = train_test_split(
X, y, test_size=0.2, stratify=y, random_state=42)

rf = RandomForestClassifier(
n_estimators=300,
max_features="sqrt",
class_weight="balanced", # menangani data tak seimbang
random_state=42)

rf.fit(X_tr, y_tr)

# Probabilitas kelas positif, bukan sekadar label
proba = rf.predict_proba(X_te)[:, 1]
print("Model training completed successfully.")
print(f"Sample probabilities: {proba[:5]}")

Model training completed successfully.
Sample probabilities: [0.         1.         0.08666667 0.31       0.00666667]


# **4.6 Implementasi: SMOTE dan Evaluasi**

In [32]:
from imblearn.over_sampling import SMOTE
from sklearn.metrics import classification_report, roc_auc_score
# SMOTE hanya pada DATA LATIH
sm = SMOTE(random_state=42)
X_res, y_res = sm.fit_resample(X_tr, y_tr)
rf.fit(X_res, y_res)
proba = rf.predict_proba(X_te)[:, 1]
pred = (proba >= 0.5).astype(int)
# Evaluasi pada data uji asli
print(classification_report(y_te, pred))
print("ROC-AUC:", roc_auc_score(y_te, proba))

              precision    recall  f1-score   support

           0       0.91      0.93      0.92        42
           1       0.96      0.94      0.95        72

    accuracy                           0.94       114
   macro avg       0.93      0.94      0.93       114
weighted avg       0.94      0.94      0.94       114

ROC-AUC: 0.9925595238095238


# **5. Aktivitas Hands-on: Customer Churn**

Langkah 1: Muat dan Eksplorasi Data

In [33]:
import pandas as pd
import os

# ambil dataset
url = "https://raw.githubusercontent.com/KimathiNewton/Telco-Customer-Churn/refs/heads/master/Datasets/telco_churn.csv"
if not os.path.exists("telco_churn.csv"):
    df = pd.read_csv(url)
    df.to_csv("telco_churn.csv", index=False)
else:
    df = pd.read_csv("telco_churn.csv")

print(f"Dataset shape: {df.shape}")
# The column name in this standard dataset is 'Churn'
print(df["Churn"].value_counts(normalize=True))

Dataset shape: (5043, 22)
Churn
False    0.440103
No       0.294923
True     0.154701
Yes      0.110274
Name: proportion, dtype: float64


catatan: memuat dataset dan menampilkan kolom churn mana pelanggan yang bertahan mana pelanggan yang kabur


---



Langkah 2: Preprocessing

In [34]:
from sklearn.model_selection import train_test_split
# TODO: encoding fitur kategorikal (mis. pd.get_dummies)
# TODO: pisahkan X (fitur) dan y (target = Churn)
X_tr, X_te, y_tr, y_te = train_test_split(
X, y, test_size=0.2, stratify=y, random_state=42)

catatan : membagi dua data, untuk data latih dan data uji dengan pembagian 20% untuk data uji


---



Langkah 3: Latih Model

In [35]:
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(
n_estimators=300, class_weight="balanced", random_state=42)
rf.fit(X_tr, y_tr)

RandomForestClassifier(class_weight='balanced', n_estimators=300,
                       random_state=42)

catatan : menjalankan algo random forest dengan pengaturan balanced dan menjalankan pelatihan model untuk mengenali ciri ciri pelanggan yg berpotensi kaburrr

---



Langkah 4: Evaluasi

In [36]:
from sklearn.metrics import classification_report, roc_auc_score
# TODO: hitung prediksi dan probabilitas
# TODO: tampilkan classification_report dan ROC-AUC

# 1. Hitung probabilitas untuk kelas positif dan tentukan prediksi
# Kita menggunakan X_te (data uji) yang sudah didefinisikan sebelumnya
proba = rf.predict_proba(X_te)[:, 1]
pred = rf.predict(X_te)

# 2. Tampilkan classification_report dan ROC-AUC
print("=== Classification Report ===")
print(classification_report(y_te, pred))

print(f"ROC-AUC Score: {roc_auc_score(y_te, proba):.4f}")

=== Classification Report ===
              precision    recall  f1-score   support

           0       0.93      0.93      0.93        42
           1       0.96      0.96      0.96        72

    accuracy                           0.95       114
   macro avg       0.94      0.94      0.94       114
weighted avg       0.95      0.95      0.95       114

ROC-AUC Score: 0.9937


catatan : hasil konerja model sangat baik akurasi mencapai 95%, angka resisi dan recall diatas 90%, artinya model bisa mengenali pelanggan yang akan kabur atau yang akan tetap tinggal, skor ROC-AUC sebsar 0.9937 artinya model memiliki kemampuan yg perfecto untuk membedakan kedua kelompok pelanggan

---



Langkah 5: Prediksi Probabilitas dan Simpulkan

In [37]:
# Hitung probabilitas churn
churn_probabilities = rf.predict_proba(X_te)[:, 1]
print(f"Contoh probabilitas churn (5 data pertama): {churn_probabilities[:5]}")

Contoh probabilitas churn (5 data pertama): [0.         1.         0.08666667 0.31       0.00666667]


**Kesimpulan singkat**


Evaluasi ini menunjukkan kinerja model yang sangat perfecto dengan akurasi menyentuh angka 95%.
Skor ROC-AUC sebesar 0.9937 menunjukan kemampuan model dalam membedakan pelanggan yang akan kabur dan yang tetap tinggal secara hampir sempurna. Kombinasi presisi dan recall yang seimbang di atas 90% menunjukan keyakinan tinggi untuk menggunakan model ini dalam strategi retensi pelanggan.
